aioauth/models.py

```diff
@@ -0,0 +1,269 @@
+
+    def check_response_type(self, response_type: ResponseType) -> bool:
+        """
+        Verifies passed ``response_type`` is part of the client's
+        ``response_types`` list.
+        """
+        return response_type in self.response_types
+
+    def get_allowed_scope(self, scope: str) -> str:
+        """
+        Returns the allowed ``scope`` given the passed ``scope``.
+
+        Note:
+            Note that the passed ``scope`` may contain multiple scopes
+            seperated by a space character.
+        """
+        if not scope:
+            return ""
+        allowed = set(self.scope.split())
+        scopes = scope_to_list(scope)
+        return list_to_scope([s for s in scopes if s in allowed])
+
+    def check_scope(self, scope: str) -> bool:
+        """Checks if passed ``scope`` is allowed for the client."""
+        allowed_scope = self.get_allowed_scope(scope)
+        return not (set(scope_to_list(scope)) - set(scope_to_list(allowed_scope)))
+
+
+class AuthorizationCode(SimpleNamespace):
+    code: str
+    """
+    Authorization code that the client previously received from the
+    authorization server.
+    """
+
+    client_id: str
+    """
+    Public identifier for the client. It must also be unique across all
+    clients that the authorization server handles.
+    """
+
+    redirect_uri: str
+    """
+    After a user successfully authorizes an application, the
+    authorization server will redirect the user back to the application
+    with either an authorization code or access token in the URL.
+    Because the redirect URL will contain sensitive information, it is
+    critical that the service doesn’t redirect the user to arbitrary
+    locations.
+    """
+
+    response_type: ResponseType
+    """A string containing the type of the response expected."""
+
+    scope: str
+    """
+    Scope is a mechanism that limit an application's access to a user's
+    account. An application can request one or more scopes, this
+    information is then presented to the user in the consent screen, and
+    the access token issued to the application will be limited to the
+    scopes granted.
+    """
+
+    auth_time: int
+    """
+    JSON Web Token Claim indicating the time when the authentication
+    occurred.
+    """
+
+    code_challenge: Optional[str] = None
+    """
+    Only used when `RFC 7636 <tools.ietf.org/html/rfc7636>`_,
+    Proof Key for Code Exchange, is used.
+
+    PKCE works by having the app generate a random value at the
+    beginning of the flow called a Code Verifier. The app hashes the
+    Code Verifier and the result is called the Code Challenge. The app
+    then kicks off the flow in the normal way, except that it includes
+    the Code Challenge in the query string for the request to the
+    Authorization Server.
+    """
+
+    code_challenge_method: Optional[CodeChallengeMethod] = None
+    """
+    Only used when `RFC 7636 <tools.ietf.org/html/rfc7636>`_,
```

Comment:
I think the link here must be also with http(s) prefix: https://synchronizing.github.io/aioauth/sections/documentation/models.html#aioauth.models.AuthorizationCode.code_challenge_method

Reply:
Sphinx wasn't playing nice at the time, and adding `https://` was (surprisingly enough) breaking the hyperlink. Without it seemed to work, but I'll go ahead and leave this review open to go over this before finalizing things - I'll try to make `https` work.